portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,35 +14,98 @@
-      "A strawberry is not an actual berry", 
-      "Kiwis contain more Vitamin C than oranges", 
-      "The pineapple is actually a berry");
+  private Map<String, String> oldFacts = new HashMap<String, String>();
+  private static final String FIRST_NAME_KEY = "first-name";
+  private static final String LAST_NAME_KEY = "last-name";
+  private static final String FUN_FACT_KEY = "user-fun-fact";
+  private static final String NO_NAME = "Anonymous";
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String fruitFactsJson = convertToJsonUsingGson(FRUIT_FACTS);
+    String oldFactsJson = convertToJsonUsingGson(oldFacts);
 
     // Send the JSON as the response
     response.setContentType("application/json;");
-    response.getWriter().println(fruitFactsJson);
+    response.getWriter().println(oldFactsJson);
   }
 
-  private String convertToJsonUsingGson(ImmutableList<String> fruitFacts) {
+  private static String convertToJsonUsingGson(Map<String, String> oldFacts) {
     Gson gson = new Gson();
-    String json = gson.toJson(fruitFacts);
+    String json = gson.toJson(oldFacts);
     return json;
   }
-}
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    String userFirstName = getParameter(request, /*content=*/ FIRST_NAME_KEY, /*defaultValue=*/ "");
+    String userLastName = getParameter(request, /*content=*/ LAST_NAME_KEY, /*defaultValue=*/ "");
+    String userFunFact = getParameter(request, /*content=*/ FUN_FACT_KEY, /*defaultValue=*/ "");
+
+    // No need to change oldFacts if there is no new fact
+    if (userFunFact.isEmpty()) {
+      response.sendRedirect("/index.html");
+      return;
+    }
+
+    // Define key and value pair to add to oldFacts
+    String factKey = getFactKey(userFirstName, userLastName);
+    String factValue = getFactValue(oldFacts, userFunFact, factKey);
+    
+    // Store user's input to access later
+    oldFacts.put(factKey, factValue);
+
+    // Redirect back to the HTML page.
+    response.sendRedirect("/index.html");
+    response.getWriter().println(factKey + ": " + factValue);
+  }
+
+  private static String getFactKey(String firstName, String lastName) {
+    String factKey;
+    if (firstName.isEmpty() && lastName.isEmpty()) {
+      factKey = NO_NAME;
+    }
+    else if (firstName.isEmpty()) {
+      factKey = lastName;
+    }
+    else if (lastName.isEmpty()) {
+      factKey = firstName;
+    }
+    else {
+      factKey = firstName + " " + lastName;
+    }
+    return factKey;
+  }
+
+  private static String getFactValue(Map<String, String> oldFacts, String funFact, String factKey) {
+    String factValue;
+    if (oldFacts.containsKey(factKey)) {
+      factValue = oldFacts.get(factKey) + ", " + funFact;
+      oldFacts.remove(factKey);
+    }
+    else {
+      factValue = funFact;
+    }
+    return factValue;
+  }
```

Comment:
Ah, I really like having this be a helper method too. But actually, you don't want to make this method static. The problem is since you're passing in oldFacts, it's actually a copy and not the actual object. Which means when you call .remove in this method, it won't modify the "real" oldFacts defined in the class.

The most straightforward solution is to just remove "static" and then you don't need to (you shouldn't) pass in the map to this method; instead you can just call it directly (since non-static methods have access to instance variables like the "real" oldFacts).

Alternatively, you could keep this method and just remove the oldFacts.remove(factKey) line. Since you always put the (factKey, factValue) pair back into the map anyways in doPost(), this isn't needed (it'll just replace it if it already exists).

Either solution works; they both have pros/cons. The first one is good but relies on non-static helper methods which can sometimes be less easy to understand because you generally want all non-static operations together so you can see exactly what's being modified. But not a huge deal. The second doesn't have this problem but copying a map can be a bit expensive. But that's just a slight optimization. Personally, I'm leaning towards the second approach, but either way works!

Reply:
Oh, I see, that makes sense. I just removed the .remove(factKey) portion. Thanks for bringing it up!